ext/ffi/lib.rs

```diff
@@ -1397,6 +1397,12 @@ struct CallbackInfo {
   pub callback: NonNull<v8::Function>,
   pub context: NonNull<v8::Context>,
   pub isolate: *mut v8::Isolate,
+  pub state: Arc<Mutex<SharedState>>,
+}
+
+struct SharedState {
+  refed: bool,
+  waker: Option<Waker>,
 }
```

Comment:
Why do we need a Mutex? Waker is Send + Sync. `refed` can be part of CallbackInfo

Reply:
> Why do we need a Mutex? Waker is Send + Sync. `refed` can be part of CallbackInfo

I have no idea :D This is quite directly copied over from here: https://rust-lang.github.io/async-book/02_execution/03_wakeups.html

Reply:
I tried removing the whole `SharedState` and moving both `refed` and `waker` to the `CallbackInfo` struct but moving even just the `refed` bool caused tests to hang. Moving `waker` along with it did not fix the situation.

Reply:
After making the resolving of the CallbackInfo Future happen only through cancel I was now able to remove the `Arc<Mutex<T>>` part and just park the waker inside of `CallbackInfo`. Foreign threads will call `waker.wake_by_ref()` to avoid unnecessary (?) clones.

Locally tests still pass so this seems fine. Maybe.